dask_sql/physical/rel/logical/limit.py

```diff
@@ -59,11 +59,16 @@ def _apply_offset(self, df: dd.DataFrame, offset: int, end: int) -> dd.DataFrame
         function, which is not possible with normal "map_partitions".
         """
         if not offset:
-            # We do a (hopefully) very quick check: if the first partition
-            # is already enough, we will just use this
-            first_partition_length = len(df.partitions[0])
-            if first_partition_length >= end:
-                return df.head(end, compute=False)
+            # compute the number of partitions that need to be persisted in head
+            npartitions = 0
+            nrows = 0
+            for partition in df.partitions:
+                npartitions += 1
+                nrows += len(partition)
```

Comment:
I might be wrong here but iiuc `len(partition)` is an eager operation, assuming we have a join followed by a `limit`, would dask compute the join `n` number of times until it found a non empty partition in this case?

Reply:
Hmm that's a good point, and perhaps why there was originally a call to `persist` preceding this check that we removed in https://github.com/dask-contrib/dask-sql/pull/387 (to avoid a duplicate computation in the case that `head()` doesn't suffice).

I wonder how bad it would be if we simply made a call to `df.head(end, npartitions=-1, compute=False)` here? Obviously that would trigger computations on partitions that might not be needed, but it would also push all the persisting to the `compute()` call instead of doing it here.